qualtran/bloqs/cryptography/ecc/ec_add.py

```diff
@@ -729,9 +748,31 @@ def build_composite_bloq(
             z4_split[i] = ctrls[1]
         z4 = bb.join(z4_split, dtype=QMontgomeryUInt(self.n))
         lam = bb.join(lam_split, dtype=QMontgomeryUInt(self.n))
-        # TODO(https://github.com/quantumlib/Qualtran/issues/1461): Fix bug in circuit where lambda
-        # is not set to 0 before being freed.
-        bb.add(Free(QMontgomeryUInt(self.n), dirty=True), reg=lam)
+
+        # If the denominator of lambda is 0, lam = lam_r so we clear lam with lam_r.
+        ancilla = bb.allocate()
+        x_split = bb.split(x)
+        x_split, ancilla = bb.add(
+            MultiControlX(cvs=[0] * int(self.n)), controls=x_split, target=ancilla
+        )
+        lam_r_split = bb.split(lam_r)
+        lam_split = bb.split(lam)
+        for i in range(int(self.n)):
+            ctrls = [ctrl, ancilla, lam_r_split[i]]
+            ctrls, lam_split[i] = bb.add(
+                MultiControlX(cvs=[1, 1, 1]), controls=ctrls, target=lam_split[i]
+            )
+            ctrl = ctrls[0]
+            ancilla = ctrls[1]
+            lam_r_split[i] = ctrls[2]
+        lam_r = bb.join(lam_r_split, dtype=QMontgomeryUInt(self.n))
+        lam = bb.join(lam_split, dtype=QMontgomeryUInt(self.n))
+        x_split, ancilla = bb.add(
+            MultiControlX(cvs=[0] * int(self.n)), controls=x_split, target=ancilla
+        )
+        x = bb.join(x_split, dtype=QMontgomeryUInt(self.n))
+        bb.free(ancilla)
```

Comment:
```suggestion
        clear_lam = (
            Xor(QMontgomeryUInt(self.n))
            .controlled(CtrlSpec(qdtypes=QMontgomeryUInt(self.n), cvs=0))
            .controlled()
        )
        ctrl, x, lam_r, lam = bb.add(clear_lam, ctrl1=ctrl, ctrl2=x, x=lam_r, y=lam)
```

You can also factor the `clear_lam` bloq into a class property, so it can also be used in the call graph.

Reply:
I tried replacing the code as suggested with the clear_lam bloq, but it gives me an error: qualtran._infra.composite_bloq.BloqError: During finalization, {Soquet(binst=BloqInstance(bloq=ControlledViaAnd(subbloq=Xor(dtype=QMontgomeryUInt(bitsize=8, modulus=None)), ctrl_spec=CtrlSpec(qdtypes=(QBit(), QMontgomeryUInt(bitsize=8, modulus=None)), cvs=(array(1), array(0)))), i=14), reg=Register(name='y', dtype=QMontgomeryUInt(bitsize=8, modulus=None), _shape=(), side=<Side.THRU: 3>), idx=())} Soquets were not used.

Do you know what's going wrong here? I'm not really understanding why it's complaining

Reply:
Hey, could you tell me which test raises this error? I replaced it locally and the tests seem to all pass.

Reply:
Did you accidentally remove the next line that frees `lam`? That may have caused the above issue